README.md
Cliprompt
==============

[![Gem Version](https://badge.fury.io/rb/cliprompt.png)](http://rubygems.org/gems/cliprompt)
[![Downloads](http://img.shields.io/gem/dt/cliprompt.svg)](https://rubygems.org/gems/cliprompt)
[![Build Status](https://travis-ci.org/mose/cliprompt.png?branch=master)](https://travis-ci.org/mose/cliprompt)
[![Coverage Status](https://coveralls.io/repos/mose/cliprompt/badge.png?branch=master)](https://coveralls.io/r/mose/cliprompt?branch=master)
[![Dependency Status](https://gemnasium.com/mose/cliprompt.svg)](https://gemnasium.com/mose/cliprompt)
[![Code Climate](https://codeclimate.com/github/mose/cliprompt.png)](https://codeclimate.com/github/mose/cliprompt)
[![Inch CI](https://inch-ci.org/github/mose/cliprompt.svg)](https://inch-ci.org/github/mose/cliprompt)

This library provides a simple DSL for managing user interaction in a CLI application. Still under development, but usable already. Check the [Changelog](https://github.com/mose/cliprompt/blob/master/CHANGELOG.md) to see what is in already.

Features
----------
- manages questions, choices, default values, yes/no values (done)
- makes possible to have env vars set for defaults (done)

Usage
----------

    # add in your Gemfile
    gem 'cliprompt'

Then in your code

```rb
require 'cliprompt'

class Myclass
  extend Cliprompt

  def initialize
    @url = ask "What is the url of Myclass?"
    @ssl = ask "Is it using SSL?", 'y/N'
    @age = ask "What is the age of the captain?", [22,33,=44,55]
    # if you set set ENV['SOMEVAR'], it will override the answer and won't display the question
    # in such case you still can use normal options and nev var will be validated against it (for boolean or choices setup)
    @age = guess 'SOMEVAR', "What is the age of the captain?", [22,33,=44,55]
  end
end
```

Check [example.rb](https://github.com/mose/cliprompt/blob/master/example.rb) for various possible combinations, or run `rspec -f d` after all that's also what tests are for.

Contributing
-----------------

1. Fork it ( https://github.com/[my-github-username]/cliprompt/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

License
----------

Copyrite (c) 2014 - mose - Distributed under MIT license


CHANGELOG.md
Cliprompt Changelog
=====================

v0.1.2 - 2015-11-17
-------------------
- fixed buggy packaging

v0.1.1 - 2015-11-16
-------------------
- adjusted more for rspec3
- better gemspec declaration

v0.1.0 - 2014-05-10
-------------------
- fix tests for rspec 3 compliance

v0.0.6 - 2014-08-10
--------------------
- fix validation of list type to be a number

v0.0.5 - 2014-05-27
--------------------
- fix to one space after question that was 2 spaces when no default is provided
- when choices are a list, show the default in the question, in more than in the list
- clean some ruby warnings (running with --debug helps)

v0.0.4 - 2014-05-24
--------------------
- add an option `aslist` for we can decide if we want it as a list or not
- add a way to display big list, when choices are more than 5, make a list out of them

v0.0.3 - 2014-05-20
--------------------
- relax the boolean check so 1 and 0 can be used for yes and no (especially for env vars)
- added a `guess` method to have env var override question if set

v0.0.2 - 2014-05-20
--------------------
- fix case when numbers are used in choices
- added some test coverage for further evolution easiness

v0.0.1 - 2014-05-19
--------------------
- first draft


spec/spec_helper.rb
$LOAD_PATH << File.expand_path('../../lib', __FILE__)
require 'rubygems'
require 'bundler'
require 'rspec/given'

if ENV['COV']
  require 'simplecov'
  SimpleCov.profiles.define :hpcli do
    add_filter '/vendor/'
    add_filter '/spec/'
  end
  SimpleCov.start :hpcli
else
  require "codeclimate-test-reporter"
  CodeClimate::TestReporter.start
end

RSpec.configure do |config|
  config.mock_with :rspec
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/lib/cliprompt_spec.rb
# encoding: utf-8

require 'spec_helper'
require 'cliprompt'
require 'cliprompt/optionset'

describe Cliprompt do
  Given(:input) { StringIO.new }
  Given(:output) { StringIO.new }
  Given { subject.setio(input, output) }

  describe '.ask' do
    Given(:question) { 'wazza?' }
    Given(:answer) { 'xxx' }
    context 'without default,' do
      When(:args) { }
      When { allow(input).to receive(:gets).and_return answer }
      Then { expect(subject.ask(question, args)).to eq answer }
      And  { expect(output.string).to eq "#{question} " }
    end
    context 'with a default,' do
      When(:default) { 'ooo' }
      When(:args) { { default: default } }
      When { allow(input).to receive(:gets).and_return answer }
      Then { expect(subject.ask(question, args)).to eq answer }
      And  { expect(output.string).to eq "#{question} [#{default}] " }
    end
    context 'with an optionset,' do
      When(:args) { Cliprompt::Optionset.new() }
      When { allow(input).to receive(:gets).and_return answer }
      Then { expect(subject.ask(question, args)).to eq answer }
      And  { expect(output.string).to eq "#{question} " }
    end
    context 'with choices requested as a list,' do
      When(:args) { { choices: ['aaa', 'bbb', 'ccc'], aslist: true } }
      When { allow(input).to receive(:gets).and_return '1' }
      Then { expect(subject.ask(question, args)).to eq 'bbb' }
      And  { expect(output.string).to eq "#{question} \n  0   aaa\n  1   bbb\n  2   ccc\n#{Cliprompt::MSG_CHOSE_A_NUMBER} " }
    end
    context 'with choices requested as a list with a default,' do
      When(:args) { { choices: ['=aaa', 'bbb', 'ccc'], aslist: true } }
      When { allow(input).to receive(:gets).and_return '' }
      Then { expect(subject.ask(question, args)).to eq 'aaa' }
      And  { expect(output.string).to eq "#{question} \n> 0   aaa\n  1   bbb\n  2   ccc\n#{Cliprompt::MSG_CHOSE_A_NUMBER} [0] " }
    end
    context 'with choices requested as a non-list despite there are many choices,' do
      When(:args) { { choices: %w(aaa bbb ccc ddd eee fff), aslist: false } }
      When { allow(input).to receive(:gets).and_return 'bbb' }
      Then { expect(subject.ask(question, args)).to eq 'bbb' }
      And  { expect(output.string).to eq "#{question} (aaa / bbb / ccc / ddd / eee / fff) " }
    end
    context 'with a big list of choices,' do
      When(:args) { %w(aaa bbb ccc ddd eee fff) }
      When { allow(input).to receive(:gets).and_return '5' }
      Then { expect(subject.ask(question, args)).to eq 'fff' }
      And  { expect(output.string).to eq "#{question} \n  0   aaa\n  1   bbb\n  2   ccc\n  3   ddd\n  4   eee\n  5   fff\n#{Cliprompt::MSG_CHOSE_A_NUMBER} " }
    end
  end

  describe '.guess' do
    Given(:question) { 'wazza?' }
    Given(:env_var) { 'SOMEVAR' }
    Given(:args) { }
    context 'when env var is provided,' do
      When(:env_value) { 'xxx' }
      When { ENV[env_var] = env_value }
      Then { expect(subject.guess(env_var, question, args)).to eq env_value }
      And  { ENV.delete(env_var) }
    end
    context 'when env var is not provided,' do
      When(:answer) { 'ooo' }
      When { allow(input).to receive(:gets).and_return answer }
      Then { expect(subject.guess(env_var, question, args)).to eq answer }
      And  { expect(output.string).to eq "#{question} " }
    end
  end

  describe '.say' do
    Given(:msg) { "hah" }
    When { subject.say msg }
    Then { expect(output.string).to eq "#{msg}\n" }
  end

  describe '.shout' do
    Given(:msg) { "hah" }
    When { subject.shout msg }
    Then { expect(output.string).to eq "#{Paint[msg, :bold, :red ]}\n" }
  end

end


spec/lib/cliprompt/optionset_spec.rb
# encoding: utf-8

require 'spec_helper'
require 'cliprompt'
require 'cliprompt/optionset'

describe Cliprompt::Optionset do

  describe '.new' do
    context "when an unknown objkect is passed as option," do
      When(:options) { Object.new }
      Then { expect{ Cliprompt::Optionset.new(options) }.to raise_error(Cliprompt::OptionException) }
    end
  end

  describe '.parse_array' do

    context "when there is no default (#{['xxx', 'yyy', 'zzz'].to_s})," do
      Given(:options) { ['xxx', 'yyy', 'zzz'] }
      Given(:set) { Cliprompt::Optionset.new(options) }
      When(:choices) { set.choices }
      When(:default) { set.default }
      When(:display) { set.display }
      Then { expect(choices).to eq options }
      Then { expect(default).to be_falsey }
      Then { expect(set.display).to eq '(xxx / yyy / zzz) ' }
    end

    context "when there is a default specified (#{['xxx', '=yyy', 'zzz'].to_s})," do
      Given(:options) { ['xxx', 'yyy', 'zzz'] }
      Given(:options_with_default) { ['xxx', '=yyy', 'zzz'] }
      Given(:set) { Cliprompt::Optionset.new(options_with_default) }
      When(:choices) { set.choices }
      When(:default) { set.default }
      When(:display) { set.display }
      Then { expect(choices).to eq options }
      Then { expect(default).to eq 'yyy' }
      Then { expect(display).to eq '(xxx / yyy / zzz) [yyy] ' }
    end

    context "when there is a mixed numeric and string choices (#{[22, 'yyy', 'zzz'].to_s})," do
      Given(:options) { [22, 'yyy', 'zzz'] }
      Given(:set) { Cliprompt::Optionset.new(options) }
      When(:choices) { set.choices }
      When(:default) { set.default }
      When(:display) { set.display }
      Then { expect(choices).to eq options.map(&:to_s) }
      Then { expect(default).to be_falsey }
      Then { expect(display).to eq '(22 / yyy / zzz) ' }
    end
  end

  describe '.parse_hash' do

    context 'when there is choices and default,' do
      context "when using sym keys (#{{ default: 'xxx', choices: ['xxx', 'yyy', 'zzz'] }.to_s})," do
        Given(:options) { { default: 'xxx', choices: ['xxx', 'yyy', 'zzz'] } }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:choices) { set.choices }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(choices).to eq ['xxx', 'yyy', 'zzz'] }
        Then { expect(default).to eq 'xxx' }
        Then { expect(display).to eq '(xxx / yyy / zzz) [xxx] ' }
      end
      context "when using string keys (#{{ 'default' => 'xxx', 'choices' => ['xxx', 'yyy', 'zzz'] }.to_s})," do
        Given(:options) { { 'default' => 'xxx', 'choices' => ['xxx', 'yyy', 'zzz'] } }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:choices) { set.choices }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(choices).to eq ['xxx', 'yyy', 'zzz'] }
        Then { expect(default).to eq 'xxx' }
        Then { expect(display).to eq '(xxx / yyy / zzz) [xxx] ' }
      end
    end

    context 'when there is only boolean,' do
      context 'when no default is given, default booean is Y,' do
        Given(:options) { { boolean: true } }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        When(:boolean) { set.boolean }
        Then { expect(default).to be_truthy }
        Then { expect(boolean).to be_truthy }
        Then { expect(display).to eq '[Y/n] ' }
      end
      context 'when default is given as true, default booean is Y,' do
        Given(:options) { { boolean: true, default: true } }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        When(:boolean) { set.boolean }
        Then { expect(default).to be_truthy }
        Then { expect(boolean).to be_truthy }
        Then { expect(display).to eq '[Y/n] ' }
      end
       context 'when default is given as false, default boolean is N,' do
        Given(:options) { { boolean: true, default: false } }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        When(:boolean) { set.boolean }
        Then { expect(default).to be_falsey }
        Then { expect(boolean).to be_truthy }
        Then { expect(display).to eq '[y/N] ' }
      end
   end
  end

  describe '.parse_fixnum' do
    Given(:options) { 42 }
    Given(:set) { Cliprompt::Optionset.new(options) }
    When(:default) { set.default }
    Then { expect(default).to eq options.to_s }
  end

  describe '.parse_fixnum' do
    Given(:options) { 3.14 }
    Given(:set) { Cliprompt::Optionset.new(options) }
    When(:default) { set.default }
    Then { expect(default).to eq options.to_s }
  end

  describe '.parse_string' do

    context 'when a random string is passed, it is the default,' do
      Given(:options) { 'something' }
      Given(:set) { Cliprompt::Optionset.new(options) }
      When(:default) { set.default }
      When(:display) { set.display }
      Then { expect(default).to eq options }
      Then { expect(display).to eq '[something] ' }
    end

    context 'when a "yesno" kind of string is passed,' do
      context 'when using yesno,' do
        Given(:options) { 'yesno' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        When(:boolean) { set.boolean }
        Then { expect(default).to be_truthy }
        Then { expect(boolean).to be_truthy }
        Then { expect(display).to eq '[Y/n] ' }
      end
      context 'when using yn,' do
        Given(:options) { 'yn' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(default).to be_truthy }
        Then { expect(display).to eq '[Y/n] ' }
      end
      context 'when using YN,' do
        Given(:options) { 'YN' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(default).to be_truthy }
        Then { expect(display).to eq '[Y/n] ' }
      end
      context 'when using yesNo,' do
        Given(:options) { 'yesNo' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(default).to be_falsey }
        Then { expect(display).to eq '[y/N] ' }
      end
      context 'when using yN,' do
        Given(:options) { 'yesNo' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(default).to be_falsey }
        Then { expect(display).to eq '[y/N] ' }
      end
      context 'when using y/N,' do
        Given(:options) { 'yesNo' }
        Given(:set) { Cliprompt::Optionset.new(options) }
        When(:default) { set.default }
        When(:display) { set.display }
        Then { expect(default).to be_falsey }
        Then { expect(display).to eq '[y/N] ' }
      end
    end

  end

  describe '.validate' do
    Given(:question) { 'so what?' }
    context 'when enter key is hit,' do
      Given(:input) { '' }
      Given(:set) { Cliprompt::Optionset.new() }
      When { allow(set).to receive(:check_default).with(question) }
      Then { expect(set).to receive(:check_default).with(question) }
      And  { expect{ set.validate(question, input) }.not_to raise_error }
    end
    context 'when it is a boolean,' do
      Given(:input) { 'x' }
      Given(:set) { Cliprompt::Optionset.new(boolean: true) }
      When { allow(set).to receive(:check_boolean).with(question, input) }
      Then { expect(set).to receive(:check_boolean).with(question, input) }
      And  { expect{ set.validate(question, input) }.not_to raise_error }
    end
    context 'when it is a choice list,' do
      Given(:input) { 'x' }
      Given(:set) { Cliprompt::Optionset.new(choices: [1,2,3]) }
      When { allow(set).to receive(:check_choices).with(question, input) }
      Then { expect(set).to receive(:check_choices).with(question, input) }
      And  { expect{ set.validate(question, input) }.not_to raise_error }
    end
    context 'when it is not a mandatory question, a boolean nor a choice list' do
      Given(:input) { 'something' }
      Given(:set) { Cliprompt::Optionset.new() }
      Then { expect(set.validate(question, input)).to eq input }
    end
  end

  describe '.check_default' do
    Given(:question) { 'so what?' }
    Given(:msg) { Cliprompt::MSG_MANDATORY_TEXT }
    context 'when there is no default set,' do
      Given(:set) { Cliprompt::Optionset.new() }
      When { allow(set).to receive(:ask_again).with(question, msg) }
      Then { expect(set).to receive(:ask_again).with(question, msg) }
      And  { expect{ set.check_default(question) }.not_to raise_error }
     end
    context 'when there is a default set,' do
      Given(:default) { 'x' }
      Given(:set) { Cliprompt::Optionset.new(default: default) }
      When(:response) { set.check_default(question) }
      Then { expect(response).to eq default }
     end
  end


  describe '.check_boolean' do
    Given(:question) { 'so what?' }
    Given(:msg) { Cliprompt::MSG_YES_OR_NO }
    Given(:set) { Cliprompt::Optionset.new(boolean: true) }
    context 'when a non yes-no answer is given,' do
      When(:input) { 'xxx' }
      When { allow(set).to receive(:ask_again).with(question, msg) }
      Then { expect(set).to receive(:ask_again).with(question, msg) }
      And  { expect{ set.check_boolean(question, input) }.not_to raise_error }
     end
    context 'when a no answer is given,' do
      When(:input) { 'no' }
      Then { expect(set.check_boolean(question, input)).to be_falsey }
    end
    context 'when a N answer is given,' do
      When(:input) { 'N' }
      Then { expect(set.check_boolean(question, input)).to be_falsey }
    end
    context 'when a yes answer is given,' do
      When(:input) { 'yes' }
      Then { expect(set.check_boolean(question, input)).to be_truthy }
    end
    context 'when a Y answer is given,' do
      When(:input) { 'Y' }
      Then { expect(set.check_boolean(question, input)).to be_truthy }
    end
  end

  describe '.check_choices' do
    Given(:question) { 'so what?' }
    Given(:msg) { Cliprompt::MSG_CHOSE_IN_LIST }
    Given(:choices) { %w(a b c) }
    Given(:set) { Cliprompt::Optionset.new(choices: choices) }
    context 'when answer is not in choices list,' do
      When(:input) { 'x' }
      When { allow(set).to receive(:ask_again).with(question, msg) }
      Then { expect(set).to receive(:ask_again).with(question, msg) }
      And  { expect{ set.check_choices(question, input) }.not_to raise_error }
     end
    context 'when answer is in choices list,' do
      When(:input) { 'a' }
      When(:response) { set.check_choices(question, input) }
      Then { expect(response).to eq input }
     end
  end

  describe '.ask_again' do
    Given(:question) { 'so what?' }
    Given(:msg) { 'heho gimme something' }
    Given(:set) { Cliprompt::Optionset.new() }
    When { allow(Cliprompt).to receive(:shout).with(msg) }
    When { allow(Cliprompt).to receive(:ask).with(question, set) }
    Then { expect(Cliprompt).to receive(:shout).with(msg) }
    And  { expect(Cliprompt).to receive(:ask).with(question, set) }
    And  { expect{ set.ask_again(question, msg)}.not_to raise_error }
  end


end


lib/cliprompt.rb
require "paint"

require "cliprompt/version"
require "cliprompt/optionset"

module Cliprompt

  module_function

  MSG_MANDATORY_TEXT = "Sorry you need to fill that information."
  MSG_YES_OR_NO = "You need to answer by yes, no, y, n, 1 or 0."
  MSG_CHOSE_IN_LIST = "You need to chose between the available options."
  MSG_INPUT_A_NUMBER = "You need input a number."
  MSG_CHOSE_A_NUMBER = "Choose a number:"

  # interactive setup of option
  def ask(question, *options)
    if options[0].class == Optionset
      opts = options[0]
    else
      opts = Optionset.new(*options)
    end
    output.print "#{question} #{opts.display}"
    answer = input.gets.chomp
    output.flush
    opts.validate(question, answer)
  end

  # if environment var is set, use it, otherwise, ask for a value
  def guess(env, question, *options)
    opts = Optionset.new(*options)
    if ENV[env]
      opts.validate(question, ENV[env])
    else
      ask question, opts
    end
  end

  # output a message
  # message - string to output
  def say(message)
    output.puts message
  end

  # paints a message in bold and red and output it
  # message - string to paint
  def shout(message)
    output.puts Paint[message, :bold, :red ]
  end

  # makes possible to set input and output
  def setio(input, output)
    @@input = input
    @@output = output
  end

  # defaults to STDIN
  def input
    @@input ||= STDIN
  end

  # defaults to STDOUT
  def output
    @@output ||= STDOUT
  end

end


lib/cliprompt/optionset.rb
module Cliprompt

  class OptionException < StandardError
  end

  class Optionset

    attr_reader :choices, :default, :boolean, :envdefault, :aslist

    def initialize(options = nil)
      @choices = []
      @default = nil
      @boolean = false
      @envdefault = nil
      @aslist = false
      @type = options.class.name.downcase
      meth = "parse_#{@type}".to_sym
      if respond_to? meth
        send(meth, options)
      else
        fail OptionException, "Undefined parser ::#{meth}"
      end
    end

    # placeholder for nil type objects
    def parse_nilclass(args)
    end

    def parse_hash(args)
      @choices = args[:choices] || args['choices'] || []
      parse_array @choices
      if args[:default] == false || args['default'] == false
        @default ||= false
      else
        @default ||= args[:default] || args['default']
      end
      if args[:aslist] == false || args['aslist'] == false
        @aslist = false
      elsif args[:aslist] == true || args['aslist'] == true
        @aslist = true
      end
      @boolean = args[:boolean] || args['boolean']
      @default = true if (@boolean && @default.nil?)
      @envdefault = args[:env] || args['env']
    end

    def parse_array(args)
      @choices = args.map(&:to_s).map do |a|
        if a[0] && a[0] == '='
          @default = a[1..-1]
        else
          a
        end
      end
      @aslist = (@choices.count > 5)
    end

    def parse_fixnum(arg)
      @default = arg.to_s
    end

    def parse_float(arg)
      @default = arg.to_s
    end

    def parse_string(arg)
      if /^[yY1](es)?(\/)?[nN0](o)?/.match(arg)
        @boolean = true
        if /y(es)?(\/)?N/.match(arg)
          @default = false
        else
          @default = true
        end
      else
        @default = arg
      end
    end

    def display
      if @boolean
        display_boolean
      elsif @choices.count > 0
        if @aslist
          display_list
        else
          display_choices
        end
      else
        display_default
      end
    end

    def display_boolean
      @default ? "[Y/n] " : "[y/N] "
    end

    def display_list
      back = "\n"
      choices.each_with_index do |choice, i|
        if @default == choice
          back << sprintf("> %-3s %s\n", i, choice)
        else
          back << sprintf("  %-3s %s\n", i, choice)
        end
      end
      back << "#{Cliprompt::MSG_CHOSE_A_NUMBER} "
      back << display_default_index.to_s
      return back
    end

    def display_choices
      back = ''
      if @choices.count > 0
        back << "(#{@choices.join(' / ')}) "
      end
      back << display_default.to_s
      return back
    end

    def display_default
      "[#{@default}] " if @default
    end

    def display_default_index
      "[#{@choices.index(@default)}] " if @default
    end

    def validate(question, answer)
      if answer == ''
        check_default question
      elsif @boolean
        check_boolean question, answer
      elsif @choices.count > 0
        if @aslist
          check_list question, answer
        else
          check_choices question, answer
        end
      else
        answer
      end
    end

    def check_default(question)
      return ask_again(question, Cliprompt::MSG_MANDATORY_TEXT) if @default.nil?
      @default
    end

    def check_boolean(question, answer)
      return ask_again(question, Cliprompt::MSG_YES_OR_NO) unless /^([yY1](es)?|[nN0](o)?)$/.match(answer)
      !/^[yY1](es)?$/.match(answer).nil?
    end

    def check_choices(question, answer)
      return ask_again(question, Cliprompt::MSG_CHOSE_IN_LIST) unless @choices.include?(answer)
      answer
    end

    def check_list(question, answer)
      return ask_again(question, Cliprompt::MSG_CHOSE_IN_LIST) unless answer.to_i < @choices.count
      return ask_again(question, Cliprompt::MSG_INPUT_A_NUMBER) unless /[0-9]+/.match(answer)
      @choices[answer.to_i]
    end

    def ask_again(question, msg)
      Cliprompt.shout msg
      Cliprompt.ask question, self
    end

  end
end


example.rb
require_relative 'lib/cliprompt'

class Myclass
  include Cliprompt

  def initialize
  end

  def askit
    puts '-------------------'
    puts 'numeric default'
    show "what is your age?", 42
    show "what is your size?", 6.2
    puts '-------------------'
    puts 'Free form'
    show "This simply ask for a simple form mandatory thing?"
    show "This simply ask for a simple form mandatory thing?", 'with a default'
    show "This simply ask for a simple form mandatory thing?", default: 'with a default again'
    puts '-------------------'
    puts 'yes/no'
    show 'a boolean?', 'y/N'
    show 'a boolean?', 'yN'
    show 'a boolean?', 'yesno'
    show 'a boolean?', 'yesNo'
    show 'a boolean?', boolean: true
    show 'a boolean?', boolean: true, default: false
    puts '-------------------'
    puts 'a list of choices'
    show 'a list without default?', [22, 33, 44, '55']
    show 'a list with default?', ['22', '33', '=44', '55']
    show 'a list without default?', choices: ['22', '33', '44', '55'], default: 22
    show 'a list with default?', choices: ['22', 33, '=44', '55']
  end

  def guessit
    puts '-------------------'
    showguess 'SOMEVAR', "what is your var?"
    showguess 'SOMEVAR', "what is your var?", 42
    showguess 'SOMEVAR', "what is your var?", boolean: true
    showguess 'SOMEVAR', "what is your var?", [12, 'aa']
  end

  def biglist
    puts '-------------------'
    puts 'a long list of choices'
    show 'a list with default?', [22, 33, 44, '=55', 66, 77, 'something else']
    show 'a list without default?', [22, 33, 44, '55', 66, 77, 'something else']
    show 'a big list of choices but not in list?', choices: [22, 33, 44, '=55', 66, 77, 'something else'], aslist: false
    show 'a small list of choices but we want it in list?', choices: [22, 33, '=44'], aslist: true
  end

  def show(*args)
    it = ask *args
    puts "-- returned #{it.inspect}"
    puts
  end

  def showguess(*args)
    it = guess *args
    puts "-- returned #{it.inspect}"
    puts
  end

end

m = Myclass.new
#m.guessit
#m.askit
m.biglist
